Pydy models migration

jbm950 opened the pull request:
This is a migration of pydy.models code to SymPy for testing and benchmarking purposes (as discussed in [PR #27](https://github.com/sympy/sympy_benchmarks/pull/27)). All dependencies on PyDy have been removed. Also some test code has been included. This is my first attempt at writing tests and so if something looks wrong or a test procedure is incorrect please let me know.

In the test code there are some rather large matrices that I've currently left as just a single line. Is there a better alternative or is this an acceptable method?

@sympy/mechanics

jbm950 committed:
Added models to the SymPy repository

Originally this code was a part of the PyDy repository but it was
discussed that it could be added to the SymPy repository for use with
benchmarking and testing. Some minor changes had to be made to make sure
no external dependencies were added. The functions in the code used to
return a pydy.system.System object but have been changed to return a
KanesMethod object now. Also the input order for the arguments to
.kanes_equations(...) had to be switched due to a recent change in the
api.

jbm950 committed:
Created tests for new models.py code

This script includes some simple tests for the models.py code. The tests
target different input combinations and a higher degree of freedom test
(n != 1).

moorepants commented:
For the symbolic tests you likely need to do:

```python
assert simplify(sym_expr - expected) = 0
```

For long lines, just wrap after operators to reasonable line length.

moorepants commented on sympy/physics/mechanics/models.py:
```diff
@@ -0,0 +1,227 @@
+    return kane
+
+
+def n_link_pendulum_on_cart(n=1, cart_force=True, joint_torques=False):
```
One thing that has bugged me about this one that I've been meaning to fix is that it would be cleaner if theta was wrt to the veritcal instead of the horizontal. Then when theta = 0 the pendulum is in the "nominal" position. This would be a good time to fix that. Would you be willing to change it so that theta is the rotation about z between the y axis and the pendulum axis?

jbm950 committed:
Fixed the docstrings of the models functions

Previously forgot to change the docstrings of the functions in models.py
to reflect the change of their output. This has been fixed.

jbm950 committed:
Changed the location of theta in n_pendulum_on_cart

The theta variables now represents the angle between the vertical y axis
and the pendulum axis.

jbm950 commented:
@moorepants The theta variable should now represent the angle between the y axis and the pendulum axis. If this works please let me know as I'm going to hold off on making too many changes to the tests until this has been resolved. Thanks

moorepants commented on sympy/physics/mechanics/models.py:
```diff
@@ -0,0 +1,227 @@
+
+    for i in range(n):
+        Bi = I.orientnew('B{}'.format(i), 'Axis', [q[i + 1], -I.z])
+        Bi.set_ang_vel(I, u[i + 1] * -I.z)
```
These should be about the positive Z axis, no?

jbm950 replied:
I changed it to where CW is now positive such that a positive angle theta from the y axis locates the pendulum arm in the drawing shown in the docstring. I can go back through and change this.

moorepants replied:
I'd rather change the drawing and keep everything as positive rotations to be more canonical.

moorepants commented on sympy/physics/mechanics/models.py:
```diff
@@ -0,0 +1,227 @@
+      F-> |   o --|---> x
+          |       |
+          ---------
+           o     o
```
This figure will need to be updated. A zero character looks a lot like a theta. The pendulum will need to be on the left side of the y axis and a zero character could be inserted in between the pendulum and y axis to represent theta.

jbm950 committed:
Updated n_link_pendulum... docstring drawing

The drawing in the docstring of n_link_pendulum_on_cart() should now
represent the new location of the theta variable.

jbm950 committed:
Changed theta direction in n_link_pendulum...

The direction of theta was changed from a CW positive direction to a CCW
positive direction.

jbm950 commented:
The direction of theta is now CCW+ and the drawing has been updated.

moorepants commented:
+1 to merge once tests pass.

jbm950 committed:
Updated and cleaned models tests

The function n_links_pendulum_on_cart() was changed so that theta
represents the angle with the vertical instead of the horizontal. The
tests are now updated to reflect that change. Also the long lines of
code have been wrapped as best as possible.

moorepants commented:
The models tests are timing out. Can you check how long it takes to run them? They should be super short.

moorepants commented on sympy/physics/mechanics/tests/test_models.py:
```diff
@@ -0,0 +1,200 @@
+                     m2)**2/(m0 + m1 + m2)))/(-m2**2/(m0 + m1 + m2) + m2 - (-m2*
+                     (m1 + m2)/(m0 + m1 + m2) + m2)**2/(m1 + m2 - (m1 + m2)**2/
+                     (m0 + m1 + m2)))]])
+    assert simplify(symexpr - kane1.rhs()) == Matrix([0, 0, 0, 0, 0, 0])
```
Maybe this is taking a long time. `trigsimp` might be better to explicitly call.

moorepants replied:
Two things here. symplify and rhs take a long time. Don't test against the results of rhs(), just compare the mass matrix and the forcing vector.

moorepants commented on sympy/physics/mechanics/tests/test_models.py:
```diff
@@ -0,0 +1,200 @@
+                      cos(q1)*cos(q2)) + l1*m2*(-l0*m1*cos(q1) - l0*m2*cos(q1))*
+                      cos(q2)/(m0 + m1 + m2))**2/(l0**2*m1 + l0**2*m2 - (-l0*m1*
+                      cos(q1) - l0*m2*cos(q1))**2/(m0 + m1 + m2)))]])
+    assert simplify(symexpr1 - kane1.rhs()) == Matrix([0, 0, 0, 0, 0, 0])
```
This definitely may be taking too long.

jbm950 commented:
When I run the test locally "tests finished: 4 passed, in 0.94 seconds". Why does it take more than 600x longer for the travis-ci testing? I'm going to work on changing the tests to the mass_matrix and forcing vectors this morning.

jbm950 committed:
Changed the comparisons of the sympy test_models

The tests for test_models now compare the mass_matrix and forcing
vectors of the results instead of calling rhs(). This has significantly
simplified the test code and should result in major speed improvements.

jbm950 commented:
I changed the tests to compare mass matrices and force vectors as suggested. Locally "tests finished: 4 passed, in 0.48 seconds" thus it cut the run time in about half.

moorepants commented:
Travis runs on different hardware, so it may just be slow.

moorepants commented:
Ok, this is good. Merging.
